wagtail/snippets/models.py

```diff
@@ -42,49 +52,57 @@ def get_edit_url(self, instance):
 
 
 def register_snippet(model, viewset=None):
-    if model not in SNIPPET_MODELS:
-        from wagtail.snippets.views.chooser import SnippetChooserViewSet
-        from wagtail.snippets.views.snippets import SnippetViewSet
-
-        model.get_usage = lambda obj: ReferenceIndex.get_references_to(
-            obj
-        ).group_by_source_object()
-        model.usage_url = get_snippet_usage_url
-        model.get_admin_base_path = get_admin_base_path
-        model.get_admin_url_namespace = get_admin_url_namespace
-
-        if viewset is None:
-            viewset = SnippetViewSet
-        elif isinstance(viewset, str):
-            viewset = import_string(viewset)
-
-        admin_viewset = viewset(
-            model.get_admin_url_namespace(),
-            model=model,
-            url_prefix=model.get_admin_base_path(),
-        )
-
-        chooser_viewset = SnippetChooserViewSet(
-            f"wagtailsnippetchoosers_{model._meta.app_label}_{model._meta.model_name}",
-            model=model,
-            url_prefix=f"snippets/choose/{model._meta.app_label}/{model._meta.model_name}",
-        )
-
-        viewsets.register(admin_viewset)
-        viewsets.register(chooser_viewset)
-
-        SNIPPET_MODELS.append(model)
-        SNIPPET_MODELS.sort(key=lambda x: x._meta.verbose_name)
-
-        @checks.register("panels")
-        def modeladmin_model_check(app_configs, **kwargs):
-            errors = check_panels_in_model(model, "snippets")
-            return errors
-
-        # Set up admin model forms to use AdminSnippetChooser for any ForeignKey to this model
-        register_form_field_override(
-            ForeignKey, to=model, override={"widget": AdminSnippetChooser(model=model)}
-        )
+    if DEFER_REGISTRATION:
+        # Models may not have been fully loaded yet, so defer registration until they are -
+        # add it to the list of registrations to be processed by register_deferred_snippets
+        DEFERRED_REGISTRATIONS.append((model, viewset))
+        return model
+    elif model in SNIPPET_MODELS:
+        # Do not create duplicate registrations of the same model
+        return model
+
+    from wagtail.snippets.views.chooser import SnippetChooserViewSet
+    from wagtail.snippets.views.snippets import SnippetViewSet
+
+    model.get_usage = lambda obj: ReferenceIndex.get_references_to(
+        obj
+    ).group_by_source_object()
+    model.usage_url = get_snippet_usage_url
+    model.get_admin_base_path = get_admin_base_path
+    model.get_admin_url_namespace = get_admin_url_namespace
+
+    if viewset is None:
+        viewset = SnippetViewSet
+    elif isinstance(viewset, str):
+        viewset = import_string(viewset)
+
+    admin_viewset = viewset(
+        model.get_admin_url_namespace(),
+        model=model,
+        url_prefix=model.get_admin_base_path(),
+    )
```

Comment:
@gasman It might be worth adding a comment above this line (if it is known to be where the problem occurs)

Reply:
I don't think so - that seems like the sort of comment that's liable to stagnate as time goes on. If django-filters fixes things at their end to not require fully-loaded models, and/or we add some more code here that does require them (which will be easy to do inadvertently, since there are no longer any negative consequences to it), then a comment singling out this line is going to be misleading. I think it's enough to document the fact that code within this block can expect models to be fully-loaded - we don't need to identify the bits that specifically take advantage of this property.